chia/types/weight_proof.py

```diff
@@ -100,3 +103,63 @@ class WeightProof(Streamable):
     sub_epochs: List[SubEpochData]
     sub_epoch_segments: List[SubEpochChallengeSegment]  # sampled sub epoch
     recent_chain_data: List[HeaderBlock]
+
+
+@streamable
+@dataclass(frozen=True)
+class SubSlotDataV2(Streamable):
+    proof_of_space: Optional[ProofOfSpace]  # if challenge block
+    # VDF to signage point
+    cc_signage_point: Optional[VDFProof]  # if block
+    # VDF from signage to infusion point
+    cc_infusion_point: Optional[VDFProof]  # if block
+    signage_point_index: Optional[uint8]  # if block
+    cc_slot_end: Optional[VDFProof]  # if is end of slot
+    cc_sp_vdf_output: Optional[CompressedClassgroupElement]  # if is end of slot
+    cc_ip_vdf_output: Optional[CompressedClassgroupElement]  # if is end of slot
+    cc_slot_end_output: Optional[ClassgroupElement]  # if is end of slot
+    icc_infusion_point: Optional[VDFProof]
+    icc_ip_vdf_output: Optional[CompressedClassgroupElement]
+
+    icc_slot_end: Optional[VDFProof]
+    icc_slot_end_output: Optional[ClassgroupElement]
+    cc_sp_signature: Optional[G2Element]  # if challenge block (needed for ChallengeBlockInfo)
+    ip_iters: Optional[uint64]  # if block
+    total_iters: Optional[uint128]  # if block
+
+    def is_challenge(self) -> bool:
+        if self.proof_of_space is not None:
+            return True
+        return False
+
+    def is_end_of_slot(self) -> bool:
+        if self.cc_slot_end_output is not None:
+            return True
+        return False
+
+
+@streamable
+@dataclass(frozen=True)
+class SubEpochChallengeSegmentV2(Streamable):
+    sub_epoch_n: uint32
+    sub_slot_data: List[SubSlotDataV2]
+    # only in first segment of each sub_epoch
```

Comment:
these fields are only present in the segment of each sub_epoch because we, we need them to validate all the hashes and vdfs.

we dont need them for the other parts of the subepoch because they are obtained from the previous  SubEpochChallengeSegmentV2